src/dagmc.cpp

```diff
@@ -206,15 +225,24 @@ void DAGUniverse::init_geometry()
     if (mat_str == "graveyard") {
       graveyard = vol_handle;
     }
-
     // material void checks
     if (mat_str == "void" || mat_str == "vacuum" || mat_str == "graveyard") {
       c->material_.push_back(MATERIAL_VOID);
     } else {
       if (uses_uwuw()) {
         uwuw_assign_material(vol_handle, c);
       } else {
-        legacy_assign_material(mat_str, c);
+        if (instance_material_overrides.size() > 0 and
+            instance_material_overrides.find(mat_str) !=
+              instance_material_overrides.end()) {
+
+          for (auto mat_str_instance :
```

Comment:
I think this question still stands? If I'm missing something can you point to the response or code change that resolves it?

Reply:
I might have miss understood you, if so sorry for marking this as resolved.

I added the test before the for loop.
L236-242

If the material overrides apply to multiple cells, that means that those cells shared the same initial material assignement in the same DAGMCgeometry.



I think I see where the problem is, this works to differenciate materials in case of multiple insertion of the same DAGMC Geometry, this does not aim to differenciate assignement of a single material to multiple cells of the same DAGMC Universe.

Reply:
Yeah exactly. If the multiple DAGMC cells uses the same material and those cells have multiple instances, it isn't clear to which cell the material overrides will apply. It seems as though it would be the same for both cells, resulting in the same material usage for those cells in each use of the DAGMC universe. In the end I don't think this will accomplish material differentiation as desired.

Reply:
That was the intended behaviour, 
Not to change the number of materials used in a DAGMC geometry but to replace the assigned materials each time the DAGMC is inserted again.

So if two cells share the same material assignment in the h5 file, the overriding material will be apply to both cells.

To go back to the previous loop, 
mat_A is assigned to Cell1 and Cell2 that are part of a DAGMC Geometry 
so we have:
`Cell1.material_[0] = mat_A`
`Cell2.material_[0] = mat_A`

If we insert the DAGMC geometry 3 times and we have a override element such as: `mat_A:[mat_A_1, mat_A_2, mat_A_3]`

So
`Cell1.material_[0] = mat_A_1`
`Cell1.material_[1] = mat_A_2`
`Cell1.material_[2] = mat_A_3`
`Cell2.material_[0] = mat_A_1`
`Cell2.material_[1] = mat_A_2`
`Cell2.material_[2] = mat_A_3`

I am happy to discuss if we want to have a different behaviour

Reply:
I was just looking at the `differentiate_depletable_mats` function again and it appears to me that we'd expect something more like.

```
Cell1.material_[0] = mat_A_0
Cell1.material_[1] = mat_A_1
Cell1.material_[2] = mat_A_2

Cell2.material_[0] = mat_A_3
Cell2.material_[1] = mat_A_4
Cell2.material_[2] = mat_A_5
```

Even if those two cells are in the same universe (a DAGMC universe in this case). This is because a clone of the original material is created each application of the material in a cell (including for each cell instance). 

As I see it, the problem with what is currently implemented it is that we'll get what you've described above, which is inconsistent with what would happen for CSG geometry. The question to me is, how does Cell2 know where to start reading the materials in the list provided by `material_overrides`? This is one reason I think it might make more sense to use cells/cell IDs as the keys of the `material_overrides` dictionary.

Separately, I do think that it would be more flexible and broadly applicable to use DAGMC cell IDs as the keys of the `material_override` object on `DAGMCUniverses`. This would allow someone to easily override a DAGMC material assignment outside of a depletion use case.

Reply:
> I am happy to discuss if we want to have a different behaviour, but I can see why differentiate material in a way that differentiation of a material assigned to multiple volumes of the DAGMC geometry, could be useful, but it could also be dangerous.

Can you elaborate on this a bit?

Reply:
Discussing with @nplinden we were thinking that allowing to "change" the material assignment done in the h5m file could lead to problem (i.e. differentiating material within the same geometry) while we only aimed to differentiate materials when importing multiple time the same DAGMC geomretry.


It took me a while to understand what you were saying (sorry), that as we follow the existing behaviour of `differentiate_depletable_mat`
You are absolutely right!

I see 2 possible implementation there:
- as you mention keep the override reference by name, and use the cell_Id+instance in the cloned mat and use this as key for the new assignment, 

- if it is preferable to reference the overrides per id, when assigning the materials, one could make a local copy of the overrides map and pop the elements from the override list while the are reassigned.
re using the previous example:
make a local copy of:

`override = mat_A:[mat_A_0, mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]`

```
override_copy = mat_A:[mat_A_0, mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]
[...]
Cell1.material_[0] = mat_A[0]
override_copy = mat_A:[mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]
Cell1.material_[1] = mat_A[0]
override_copy = mat_A:[mat_A_2, mat_A_3, mat_A_4, mat_A_5]
Cell1.material_[2] = mat_A[0]
override_copy = mat_A:[mat_A_3, mat_A_4, mat_A_5]
[...]
Cell1.material_[3] = mat_A[0]
override_copy = mat_A:[mat_A_4, mat_A_5]
Cell1.material_[4] = mat_A[0]
override_copy = mat_A:[mat_A_5]
Cell1.material_[5] = mat_A[0]
override_copy = mat_A:[]
[...]

```

Reply:
> Discussing with @nplinden we were thinking that allowing to "change" the material assignment done in the h5m file could lead to problem (i.e. differentiating material within the same geometry) while we only aimed to differentiate materials when importing multiple time the same DAGMC geomretry.

Ah I see. Yes, for this particular use case I can see how that might be confusing.


>  - as you mention keep the override reference by name, and use the cell_Id+instance in the cloned mat and use this as key for the new assignment,

This is close to what I had in mind. I think in the `material_overrides` dictionary each entry can use a cell ID as a key. The value of this entry can be either a list of length one (the material override applies to all cell instances) or an `Iterable` with a size equal to the number of cell instances -- this can be verified on the Python or C++ side (or both). This is how the `openmc.Cell.fill` attribute is handled and checked at runtime on the C++ side. That thought leads me back to wondering if it might be viable to add a `fill` attribute to the `DAGMCCell` object. Those 'fill' attributes can be written with the DAGMCUniverse XML node if present.

Reply:
ok so we need to enable 2 override modes :
- 1 done automatically with `differentiate_depletable_mat`, referencing `DAGMCCell.id` to a list of mats (id?) corresponding to each instance of the Cell
- 1 when user manually override materials assignment of the DAGMC h5 file, referencing  the original material_name to a list of mats (id?) corresponding to each instance of the DAGMC universe

Reply:
what we can do, if check if the `key` is in the original mat assignment and switch it to `id` search if not.

I understood you'd prefer if we reference the materials used for the override by id not by mat ?
so having something like:
`{key:[mat_id_1,...,mat_id_n}`

correct ?

Reply:
> ok so we need to enable 2 override modes :

Something like that, yeah. I think of this as two different use cases that we're trying to support with the same data structure.

I think I understand better now what your concern is regarding "dangerous" behavior. In theory it would be possible for a user to call `diff_burnable_mats` and then change the overrides dictionary entry. I'd propose two solutions for this (both involve checking the new assignments against the original assignment in the DAGMC file). 

  a. Perform the check during the `diff_burnable_mats` method as a way of verifying the updated material assignments
  b. Perform the check upon export of the DAGMC universe to XML and display a warning or raise an exception 

Personally, I like option a. I know it leaves more room for error if the user makes manual modifications after the call to `diff_burnable_mats`, but historically the design of OpenMC has favored flexibility in these cases.

Not trying to be pedantic or nit-picky. Mainly I want to make sure we're on the same page so your time isn't wasted if we aren't in agreement.

Reply:
Going back to data structures, what do you think of representing the overrides in a `DAGMCCell.fill` attribute?

Reply:
> Personally, I like option a. I know it leaves more room for error if the user makes manual modifications after the call to `diff_burnable_mats`, but historically the design of OpenMC has favored flexibility in these cases.

When talking about user manual override of the material assignment I was not referring to user manual modification after the call to `diff_burnable_mats`
but simply a user wanting to override some materials of the DAGMC geometry.
writing manually the `DAGMCUniverse.material_overrides` or using `DAGMCUniverse.add_material_override`

I think it is a nice feature to be able to "replace" material from the original DAGMC geometry by mapping the original name with the news materials (for either unique or multiple DAGMC occurrence). It is also nice that it is 100% Python so it does not require to have the `openmc.lib` enabled.

I also like the "vision" you have, when using `diff_burnable_mats` to create a mapping  between DAGMCCell.id and the materials for each instances of this cell.

IMO, we need to enable both, and find a way to pass that information correctly to the cpp.

Reply:
> Going back to data structures, what do you think of representing the overrides in a `DAGMCCell.fill` attribute?

Do you mean when using the `diff_burnable_mats` ?
because we already write the override in the `cell.fill`.
it is then used to generate the `material_override` node in the xml.

As to have `find_buranle_mats` only works with `DAGMCCell` when `openmc.lib` is enable, we could pass back the information directly to the cpp DAGMCCell. 
Or add a method to pass it.

Is that what you meant ?

Reply:
> When talking about user manual override of the material assignment I was not referring to user manual modification after the call to `diff_burnable_mats` but simply a user wanting to override some materials of the DAGMC geometry writing manually the `DAGMCUniverse.material_overrides` or using `DAGMCUniverse.add_material_override`

 > I think it is a nice feature to be able to "replace" material from the original DAGMC geometry by mapping the original name with the news materials (for either unique or multiple DAGMC occurrence). 

> IMO, we need to enable both, and find a way to pass that information correctly to the cpp.

I think we're in agreement that we can and should support both use cases of material overrides we've outlined above. My main point was that I think we can use the same override data structure (be it with a `dict` on the `DAGMCUniverse` or using the `DAGMCCell.fill` attribute).

> It is also nice that it is 100% Python so it does not require to have the `openmc.lib` enabled.

I don't quite see how we can do this without using the `openmc.lib` module at some point myself, but you have a way to accomplish this in mind, great!

Reply:
> Do you mean when using the `diff_burnable_mats` ? because we already write the override in the `cell.fill`. For `DAGMCCell`, it is then used to generate the `material_override` node in the xml.

Ah, yeah you're right. I was thinking that if the information is stored on the `DAGMCCell` objects then we're kind of duplicating it in the `material_overrides` attribute. So I was wondering if we might just populate the XML element from the `DAGMCCell` objects during export and not have an attribute on the `DAGMCUniverse`. However, I'm not sure how one would handle importing an XML file without generating `DAGMCCell` objects. I think we can look into that further down the line. Not super important at the moment.
 
> As to have `find_buranle_mats` only works with `DAGMCCell` when `openmc.lib` is enable, we could pass back the information directly to the cpp DAGMCCell. Or add a method to pass it.
> 
> Is that what you meant ?

This would be kind of nifty, but I don't think we need to worry about it at the moment.

Reply:
I added a new implementation for this.

basically in the python DAGMCUniverse before generating the xml node for the DAGMCUniverse, one update the override_materials dict with the cell.fill (if the cells have multiple instance and an `Iterable` fill, i.e. the DAGMCCell is inserted multiple times in the Geometry, and has a different fill for each instance.)

If an override_material has already been set overloading per material name, I decided to keep both.
And on loading the DAGMC geometry (C++) to prioritise the override per cell Id then  override per material name then fall back to material assignment as stated in the DAGMC file.